File: examples/python/restore.py
#!/usr/bin/env python3
"""Rejoin a split GotYouBro backup (Python 3.8+, standard library only).

Download all parts (and the .manifest.json) from Telegram into one folder, then:

    python3 restore.py ./downloads                 # → postgres-20260924T030000Z.sql.gz
    python3 restore.py ./downloads --extract       # → postgres-20260924T030000Z.sql
    python3 restore.py a.part001 a.part002 --out backup.sql.gz

With a manifest every part and the joined file are verified with SHA-256.
Without one, files named *.partNNN are joined in order.
"""

import argparse
import gzip
import hashlib
import json
import os
import re
import shutil
import sys

PART_RE = re.compile(r"^(.*)\.part(\d{3,})$")


def collect_files(inputs):
    files = []
    for item in inputs:
        if os.path.isdir(item):
            files.extend(os.path.join(item, name) for name in sorted(os.listdir(item)))
        else:
            files.append(item)
    return files


def sha256(path: str) -> str:
    digest = hashlib.sha256()
    with open(path, "rb") as fh:
        for chunk in iter(lambda: fh.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def main() -> None:
    parser = argparse.ArgumentParser(description="Rejoin a split GotYouBro backup")
    parser.add_argument("inputs", nargs="+", help="folder with the parts, or the part files")
    parser.add_argument("--out", help="output file (default: original name)")
    parser.add_argument("--extract", action="store_true", help="also gunzip the joined file")
    args = parser.parse_args()

    files = collect_files(args.inputs)
    manifest_path = next((f for f in files if f.endswith(".manifest.json")), None)
    manifest = None

    if manifest_path:
        with open(manifest_path) as fh:
            manifest = json.load(fh)
        by_name = {os.path.basename(f): f for f in files}
        folder = os.path.dirname(manifest_path)
        parts = [
            {**p, "path": by_name.get(p["file"], os.path.join(folder, p["file"]))} for p in manifest["parts"]
        ]
        out_name = manifest["file"]
    else:
        matches = [(f, PART_RE.match(os.path.basename(f))) for f in files]
        matches = sorted((m for m in matches if m[1]), key=lambda m: int(m[1].group(2)))
        if not matches:
            sys.exit("No *.partNNN files or manifest found")
        parts = [{"path": f} for f, _ in matches]
        out_name = matches[0][1].group(1)

    out = args.out or out_name
    print(f"Joining {len(parts)} part(s) → {out}")
    total = hashlib.sha256()
    with open(out, "wb") as output:
        for i, part in enumerate(parts, 1):
            path = part["path"]
            if not os.path.exists(path):
                sys.exit(f"Missing part: {os.path.basename(path)}")
            if part.get("sha256") and sha256(path) != part["sha256"]:
                sys.exit(f"Checksum mismatch in {os.path.basename(path)} — download it again")
            with open(path, "rb") as fh:
                for chunk in iter(lambda: fh.read(1024 * 1024), b""):
                    total.update(chunk)
                    output.write(chunk)
            print(f"  ✓ {i}/{len(parts)} {os.path.basename(path)}")

    if manifest:
        if total.hexdigest() != manifest["sha256"]:
            sys.exit("Checksum of the joined file does not match the manifest")
        print("Checksums verified.")

    if args.extract:
        extracted = out[:-3] if out.endswith(".gz") else f"{out}.out"
        with gzip.open(out, "rb") as src, open(extracted, "wb") as dst:
            shutil.copyfileobj(src, dst, 1024 * 1024)
        print(f"Extracted → {extracted}")
    if manifest and manifest.get("restore"):
        print(f"\nRestore with:\n  {manifest['restore'].split('&& ')[-1]}")


if __name__ == "__main__":
    main()

File: examples/node/restore.mjs
#!/usr/bin/env node
// Rejoin a split GotYouBro backup (Node.js 18+, no dependencies).
//
// Download all parts (and the .manifest.json) from Telegram into one folder, then:
//
//   node restore.mjs ./downloads                 # → postgres-20260924T030000Z.sql.gz
//   node restore.mjs ./downloads --extract       # → postgres-20260924T030000Z.sql
//   node restore.mjs a.part001 a.part002 --out backup.sql.gz
//
// With a manifest every part and the joined file are verified with SHA-256.
// Without one, files named *.partNNN are joined in order.

import { createHash } from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { pipeline } from 'node:stream/promises';
import { parseArgs } from 'node:util';
import { createGunzip } from 'node:zlib';

const PART_RE = /^(.*)\.part(\d{3,})$/;

function collectFiles(inputs) {
  const files = [];
  for (const input of inputs) {
    if (fs.statSync(input).isDirectory()) {
      for (const name of fs.readdirSync(input)) files.push(path.join(input, name));
    } else {
      files.push(input);
    }
  }
  return files;
}

async function sha256(file) {
  const hash = createHash('sha256');
  for await (const chunk of fs.createReadStream(file)) hash.update(chunk);
  return hash.digest('hex');
}

async function main() {
  const { values, positionals } = parseArgs({
    allowPositionals: true,
    options: { out: { type: 'string' }, extract: { type: 'boolean', default: false } },
  });
  if (!positionals.length) {
    console.error('Usage: node restore.mjs <folder | files...> [--out file.gz] [--extract]');
    process.exit(2);
  }

  const files = collectFiles(positionals);
  const manifestPath = files.find((f) => f.endsWith('.manifest.json'));
  let parts;
  let outName;
  let manifest;

  if (manifestPath) {
    manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf8'));
    const dir = path.dirname(manifestPath);
    parts = manifest.parts.map((p) => ({ ...p, path: files.find((f) => path.basename(f) === p.file) ?? path.join(dir, p.file) }));
    outName = manifest.file;
  } else {
    parts = files
      .map((f) => ({ path: f, match: PART_RE.exec(path.basename(f)) }))
      .filter((p) => p.match)
      .sort((a, b) => Number(a.match[2]) - Number(b.match[2]));
    if (!parts.length) throw new Error('No *.partNNN files or manifest found');
    outName = parts[0].match[1];
  }

  const out = values.out ?? outName;
  console.log(`Joining ${parts.length} part(s) → ${out}`);

  // Verify and concatenate.
  const total = createHash('sha256');
  const output = fs.createWriteStream(out);
  for (const [i, part] of parts.entries()) {
    if (!fs.existsSync(part.path)) throw new Error(`Missing part: ${path.basename(part.path)}`);
    if (part.sha256 && (await sha256(part.path)) !== part.sha256) {
      throw new Error(`Checksum mismatch in ${path.basename(part.path)} — download it again`);
    }
    for await (const chunk of fs.createReadStream(part.path)) {
      total.update(chunk);
      if (!output.write(chunk)) await new Promise((resolve) => output.once('drain', resolve));
    }
    console.log(`  ✓ ${i + 1}/${parts.length} ${path.basename(part.path)}`);
  }
  await new Promise((resolve, reject) => output.end((err) => (err ? reject(err) : resolve())));

  if (manifest && total.digest('hex') !== manifest.sha256) throw new Error('Checksum of the joined file does not match the manifest');
  if (manifest) console.log('Checksums verified.');

  if (values.extract) {
    const extracted = out.replace(/\.gz$/, '') || `${out}.out`;
    await pipeline(fs.createReadStream(out), createGunzip(), fs.createWriteStream(extracted));
    console.log(`Extracted → ${extracted}`);
  }
  if (manifest?.restore) console.log(`\nRestore with:\n  ${manifest.restore.split('&& ').pop()}`);
}

main().catch((err) => {
  console.error(`Restore failed: ${err.message}`);
  process.exit(1);
});
